### PR TITLE
[FLINK-29044] Add Transformer for DCT

### PR DESCRIPTION
**Pull request by yunfengzhou-hub**
## What is the purpose of the change

Add Transformer for DCT in Flink ML.

## Brief change log
- Added Transformer for DCT.
- Added java test/example for DCT.
- Added python source/test/example for DCT.
- The public interface is consistent with DCT in Spark ML.

## Does this pull request potentially affect one of the following parts:
- Dependencies (does it add or upgrade a dependency): (yes)
- The public API, i.e., is any changed class annotated with @public(Evolving): (no)
- Does this pull request introduce a new feature? (yes)
- If yes, how is the feature documented? (Java doc)

### CODE REVIEW

**Comment by lindong28** [comment]
Thanks for the PR. LGTM.

@zhipeng93 Do you want to take another look?

**Comment by zhipeng93** [comment]
> Thanks for the PR. LGTM.
> 
> @zhipeng93 Do you want to take another look?

Thanks for the review.  I will do a pass today.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/dct/DCT.java [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    public Table[] transform(Table... inputs) {
+        Preconditions.checkArgument(inputs.length == 1);
+
+        StreamTableEnvironment tEnv =
+                (StreamTableEnvironment) ((TableImpl) inputs[0]).getTableEnvironment();
+
+        RowTypeInfo inputTypeInfo = TableUtils.getRowTypeInfo(inputs[0].getResolvedSchema());
+        RowTypeInfo outputTypeInfo =
+                new RowTypeInfo(
+                        ArrayUtils.addAll(inputTypeInfo.getFieldTypes(), VectorTypeInfo.INSTANCE),
```
Comment (zhipeng93) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/dct/DCT.java:72]: It seems that the output is always a dense vector. How about make it `DenseVectorTypeInfo.INSTANCE`?

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/dct/DCT.java [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            this.isInverse = isInverse;
+            this.dctTransformer = null;
+            this.previousVectorSize = -1;
+        }
+
+        @Override
+        public Row map(Row row) throws Exception {
+            Vector vector = row.getFieldAs(inputCol);
+
+            if (dctTransformer == null || previousVectorSize != vector.size()) {
```
Comment (zhipeng93) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/dct/DCT.java:105]: nit: `dctTransformer == null` seems unnecessary.

File: flink-ml-lib/src/main/java/org/apache/flink/ml/feature/dct/DCTParams.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                    "inverse",
+                    "Whether to perform the inverse DCT (true) or forward DCT (false).",
+                    false);
+
+    default boolean getInverse() {
+        return get(INVERSE);
+    }
+
+    default T setInverse(boolean value) {
+        set(INVERSE, value);
```
Comment (zhipeng93) [inline flink-ml-lib/src/main/java/org/apache/flink/ml/feature/dct/DCTParams.java:43]: nit: could be `return set(INVERSE, value);`

**Review by zhipeng93 — COMMENTED** [review]
Thanks for the PR. LGTM overall. Left some minor comments below.

BTW, do you think it is a good idea to introduce `JLargeArrays` here? It needs to allocate off-heap memory when the array size is bigger than INT.MAX, which is not friendly to Flink memory management.

In this PR I am ok since the size is always smaller than INT.MAX.

**Comment by zhipeng93** [comment]
LGTM.

**Pull request merged**